### PR TITLE
Simplify TypeRef, `baseType` returns NoType, as needed for isSubtype

### PR DESCRIPTION
**Pull request by adriaanm**
I started pulling on this thread for the INDY sammy work -- ran into SI-9540, which then led me to the isSubType bug due to substitution returning ErrorType (which makes <:< say yes when it should say no, as the original type from baseType was NoType), which then led to obsessing over the complexities due to OO factoring of TypeRef.

---

TODO: 
[dropped -- not sure how to do this, but we should spend more time unit testing subtyping ] test case to show problem with subtyping that should now be fixed -- it was related to transform's call to `instantiateTypeParams` turning a NoType (because there was no base type at the given symbol) into an ErrorType (due to mismatch between type args and params), which makes subtype go from `false` to `true`...
- [x] compare ASF call counts before and after

I've a lot of stack frames to pop back up to, so I have done various manual testing, but I need to get back to the work that triggered this.

### CODE REVIEW

**Comment by adriaanm** [comment]
Looks like a test case progression:

```
 Test suite finished with 1 case failing:
fail - run/delambdafy_t6028.scala  [output differs]% scalac delambdafy_t6028.scala
% /usr/lib/jvm/java-8-openjdk-amd64/jre/bin/java \
  -Dfile.encoding=UTF-8 \
  -server \
  -XX:+AggressiveOpts \
  -XX:+UseParNewGC \
  -Xmx2G \
  -Xss1M \
  -XX:MaxPermSize=512M \
  -XX:ReservedCodeCacheSize=128M \
  -Dpartest.threads=4 \
  -classpath \
  /home/jenkins/workspace/scala-2.12.x-validate-test@2/test/files/run/delambdafy_t6028-run.obj:/home/jenkins/workspace/scala-2.12.x-validate-test@2/build/pack/lib/scala-library.jar:/home/jenkins/workspace/scala-2.12.x-validate-test@2/build/pack/lib/scala-reflect.jar:/home/jenkins/workspace/scala-2.12.x-validate-test@2/build/pack/lib/scala-compiler.jar:/home/jenkins/workspace/scala-2.12.x-validate-test@2/build/pack/lib/scalap.jar:/home/jenkins/workspace/scala-2.12.x-validate-test@2/m2repo/org/scala-lang/modules/scala-partest_2.12.0-M3/1.0.12/scala-partest_2.12.0-M3-1.0.12.jar:/home/jenkins/workspace/scala-2.12.x-validate-test@2/m2repo/com/googlecode/java-diff-utils/diffutils/1.3.0/diffutils-1.3.0.jar:/home/jenkins/workspace/scala-2.12.x-validate-test@2/m2repo/org/scala-sbt/test-interface/1.0/test-interface-1.0.jar:/home/jenkins/workspace/scala-2.12.x-validate-test@2/m2repo/org/scala-lang/modules/scala-xml_2.12.0-M3/1.0.5/scala-xml_2.12.0-M3-1.0.5.jar:/home/jenkins/workspace/scala-2.12.x-validate-test@2/m2repo/org/scala-lang/modules/scala-parser-combinators_2.12.0-M3/1.0.4/scala-parser-combinators_2.12.0-M3-1.0.4.jar:/home/jenkins/workspace/scala-2.12.x-validate-test@2/m2repo/org/scalacheck/scalacheck_2.12.0-M3/1.11.6/scalacheck_2.12.0-M3-1.11.6.jar:/home/jenkins/workspace/scala-2.12.x-validate-test@2/build/pack/lib/scala-partest-extras.jar:/home/jenkins/workspace/scala-2.12.x-validate-test@2/build/pack/lib/scala-partest-javaagent.jar:/home/jenkins/workspace/scala-2.12.x-validate-test@2/test/files/lib/annotations.jar:/home/jenkins/workspace/scala-2.12.x-validate-test@2/test/files/lib/enums.jar:/home/jenkins/workspace/scala-2.12.x-validate-test@2/test/files/lib/genericNest.jar:/home/jenkins/workspace/scala-2.12.x-validate-test@2/test/files/lib/jsoup-1.3.1.jar:/home/jenkins/workspace/scala-2.12.x-validate-test@2/test/files/lib/macro210.jar:/home/jenkins/workspace/scala-2.12.x-validate-test@2/test/files/lib/methvsfield.jar:/home/jenkins/workspace/scala-2.12.x-validate-test@2/test/files/lib/nest.jar \
  -Dfile.encoding=UTF-8 \
  -Djava.library.path=/home/jenkins/workspace/scala-2.12.x-validate-test@2/test/files/run \
  -Dpartest.output=/home/jenkins/workspace/scala-2.12.x-validate-test@2/test/files/run/delambdafy_t6028-run.obj \
  -Dpartest.lib=/home/jenkins/workspace/scala-2.12.x-validate-test@2/build/pack/lib/scala-library.jar \
  -Dpartest.reflect=/home/jenkins/workspace/scala-2.12.x-validate-test@2/build/pack/lib/scala-reflect.jar \
  -Dpartest.comp=/home/jenkins/workspace/scala-2.12.x-validate-test@2/build/pack/lib/scala-compiler.jar \
  -Dpartest.cwd=/home/jenkins/workspace/scala-2.12.x-validate-test@2/test/files/run \
  -Dpartest.test-path=/home/jenkins/workspace/scala-2.12.x-validate-test@2/test/files/run/delambdafy_t6028.scala \
  -Dpartest.testname=delambdafy_t6028 \
  -Djavacmd=/usr/lib/jvm/java-8-openjdk-amd64/jre/bin/java \
  -Djavaccmd=javac \
  -Duser.language=en \
  -Duser.country=US \
  scala.tools.nsc.MainGenericRunner \
  -usejavacp \
  Test \
  jvm > delambdafy_t6028-run.log
% diff /home/jenkins/workspace/scala-2.12.x-validate-test@2/test/files/run/delambdafy_t6028-run.log /home/jenkins/workspace/scala-2.12.x-validate-test@2/test/files/run/delambdafy_t6028.check
@@ -11,7 +11,7 @@ package <empty> {
     def foo(methodParam: String): Function0 = {
       val methodLocal: String = "";
       {
-        (() => T.this.$anonfun$1(methodParam, methodLocal))
+        (() => T.this.$anonfun$1(methodParam, methodLocal)).$asInstanceOf[Function0]()
       }
     };
     def bar(barParam: String): Object = {
@@ -21,7 +21,7 @@ package <empty> {
     def tryy(tryyParam: String): Function0 = {
       var tryyLocal: runtime.ObjectRef = scala.runtime.ObjectRef.create("");
       {
-        (() => T.this.$anonfun$2(tryyParam, tryyLocal))
+        (() => T.this.$anonfun$2(tryyParam, tryyLocal)).$asInstanceOf[Function0]()
       }
     };
     final <artifact> private[this] def $anonfun$1(methodParam$1: String, methodLocal$1: String): String = T.this.classParam.+(T.this.field()).+(methodParam$1).+(methodLocal$1);

```

**Comment by lrytz** [comment]
looks like you fixed https://issues.scala-lang.org/browse/SI-9540

**Comment by adriaanm** [comment]
[EDIT: This was on a WIP branch. The speedup is still there, but less significant.] the cleanup also yields small speedups (strap was .94 of quick's time) on my machine (locker is built from 2.12.x, no optimizer)

```
[stopwatch] [quick.library.timer: 55.512 sec]
[stopwatch] [quick.reflect.timer: 49.030 sec]
[stopwatch] [quick.compiler.timer: 1:07.699 sec]
[stopwatch] [strap.library.timer: 50.543 sec]
[stopwatch] [strap.reflect.timer: 46.968 sec]
[stopwatch] [strap.compiler.timer: 1:05.138 sec]
```

**Comment by adriaanm** [comment]
The test failure was a timeout due to behemoth overload.

**Comment by adriaanm** [comment]
/nothingtoseehere -- FYI, you can still see the original successes

**Comment by adriaanm** [comment]
review by @lrytz  & @retronym,  please (best to get more eyes on this, given this is a refactoring in the core)

**Comment by adriaanm** [comment]
I started pulling on this thread for the INDY sammy work -- ran into SI-9540, which then led me to the isSubType bug due to substitution returning ErrorType (which makes <:< say yes when it should say no, as the original type from baseType was NoType), which then led to obsessing over the complexities due to OO factoring of TypeRef.

**Comment by lrytz** [comment]
I got my :eyes: on it, they hurt a bit now :) I didn't see anything to flag, and it's a great improvement. As Jason says, it having some tests for the bugfixes would be great to have.

**Comment by retronym** [comment]
I took a look at the main part of this PR today. Haven't been able to poke any holes in it just yet. But I'll have another pass before signing off.

**Comment by retronym** [comment]
Scheduled a community build: https://scala-ci.typesafe.com/view/scala-2.12.x/job/scala-2.12.x-integrate-community-build/322/

**Comment by SethTisue** [comment]
> Scheduled a community build

the twitter-util failure looks like a transient failure to resolve dependencies

**Comment by SethTisue** [comment]
green: https://scala-ci.typesafe.com/job/scala-2.12.x-integrate-community-build/324/

**Comment by retronym** [comment]
- [x] Could you please gather before/after stats on number of calls to `isSubType` and `asSeenFrom`, to make sure that the changes to caches in `TypeRef` aren't causing additional work?

**Comment by adriaanm** [comment]
|  |  | quick ASF | strap ASF | quick substmaps | strap substmaps |
| --- | --- | --- | --- | --- | --- |
| library | parser | 276664 | 275957 | 652497 | 622386 |
| library | typer | 325897 | 325161 | 679876 | 643556 |
| library | patmat | 1028719 | 1027641 | 2367183 | 2260357 |
| library | erasure | 1048317 | 1047239 | 2539832 | 2432964 |
| library | cleanup | 1280868 | 1279790 | 2859582 | 2750405 |
|  |  |  |  |  |  |
| reflect | parser | 339418 | 337441 | 514299 | 484061 |
| reflect | typer | 530369 | 528196 | 745690 | 706088 |
| reflect | patmat | 762988 | 759696 | 1082757 | 1032918 |
| reflect | erasure | 769554 | 766262 | 1127709 | 1077832 |
| reflect | cleanup | 824371 | 821046 | 1169431 | 1119347 |
|  |  |  |  |  |  |
| compiler | parser | 667471 | 662019 | 1145647 | 1076131 |
| compiler | typer | 993941 | 987900 | 1585757 | 1501836 |
| compiler | patmat | 1125824 | 1118482 | 2005167 | 1902220 |
| compiler | erasure | 1132020 | 1124678 | 2061633 | 1958660 |
| compiler | cleanup | 1169103 | 1161758 | 2109108 | 2005895 |

**Comment by adriaanm** [comment]
so, slightly fewer ASF and substitutions after the change -- forgot to measure subtyping, will look at that asap

**Comment by smarter** [comment]
Do you have flags in the compiler to gather these statistics?

**Comment by adriaanm** [comment]
I added some for this comparison: https://github.com/adriaanm/scala/commit/831c35ba5a1ab2a55b33a65765676b438c214f09

**Comment by adriaanm** [comment]
Generating new numbers for isSubType with `rm -rf build ; g co b887fdbbda ; ant locker.done ; g co - ; ant -Dscalac.args=-Ystatistics:_ quick.bin ;  ant -Dscalac.args=-Ystatistics:_ strap.done`  (b887fdbbda is merge base for this PR + statistics commit 831c35b)

**Comment by adriaanm** [comment]
Here are the improvements for isSubtype:

```
[quick.library] isSubtypes:  2375379
[quick.reflect] isSubtypes:  6217668
[quick.compiler] isSubtypes  3959157

[strap.library] isSubtypes:  2066112
[strap.reflect] isSubtypes:  4454277
[strap.compiler] isSubtypes  3518288
```

**Comment by retronym** [comment]
At the risk of sounding paranoid, are we sure that 30% of `<:<` ops that we've shed in compiling `reflect` weren't buying us something? I wonder if we can compile a small program that exhibits a similar proportional reduction, add logging of the `safeToString` of the arguments to `isSubTypes`, and diff the log with the status quo compiler.

**Comment by adriaanm** [comment]
I think the check I added here: https://github.com/scala/scala/pull/4958/files#diff-706ee5b97253431a2f83399fea7e0786R406 explains it. As the comment says, it preempts a large number of fruitless subtype tests.

**Comment by adriaanm** [comment]
( -- time passes, math is done -- Especially since the numbers match exactly: it goes down from 12.5M to 10M :-) )

**Comment by retronym** [comment]
LGTM

File: src/reflect/scala/reflect/internal/tpe/TypeComparers.scala [hunk 1]
```diff
@@ -222,7 +222,7 @@ trait TypeComparers {
       case SingleType(pre1, sym1)     => tp2 match { case SingleType(pre2, sym2)     => equalSymsAndPrefixes(sym1, pre1, sym2, pre2)         ; case _ => false }
       case PolyType(ps1, res1)        => tp2 match { case PolyType(ps2, res2)        => equalTypeParamsAndResult(ps1, res1, ps2, res2)       ; case _ => false }
       case ExistentialType(qs1, res1) => tp2 match { case ExistentialType(qs2, res2) => equalTypeParamsAndResult(qs1, res1, qs2, res2)       ; case _ => false }
-      case ThisType(sym1)             => tp2 match { case ThisType(sym2)             => sym1 == sym2                                         ; case _ => false }
+      case ThisType(sym1)             => tp2 match { case ThisType(sym2)             => sym1 eq sym2                                         ; case _ => false }
```
Comment (smarter) [inline src/reflect/scala/reflect/internal/tpe/TypeComparers.scala:225]: Is there a difference between reference equality and value equality for symbols in scalac?

Comment (adriaanm) [inline src/reflect/scala/reflect/internal/tpe/TypeComparers.scala:225]: No, we should always use reference equality (for clarity).

**Pull request merged**